lib/ansible/modules/source_control/gitlab_service.py

```diff
@@ -0,0 +1,400 @@
+            # when saving, params does directly in the request
+            # (remember they get retrieved under the "property" key of the remote dict())
+            for k, v in params.items():
+                setattr(remote, k, v)
+            if not self._module.check_mode:
+                remote.save()
+            return remote
+        else:
+            return False
+
+    def asApiObject(self, active, params, events):
+        obj = {'active': active}
+        obj.update(self.expandEvents(events))
+        obj.update(params)
+        return obj
+
+    def expandEvents(self, events):
+        try:
+            # See https://gitlab.com/gitlab-org/gitlab-ce/issues/58321 for why it's useful to
+            # discard unsupported events before comparing
+            supported_events = RAW_SERVICES_DEFINITIONS[self.name]['_events']
+        except KeyError:
+            supported_events = None
+        return {value + '_events': bool(value in events) for value in self.HOOK_EVENTS if supported_events is None or value in supported_events}
+
+    def eventsEqual(self, obj, events):
+        return all((k in obj and obj[k] == v) for k, v in events.items())
+
+    def hasCredentials(self, attr):
+        return ('password' in attr and attr['password']) or ('token' in attr and attr['token']) or ('api_key' in attr and attr['api_key'])
+
+    def equals(self, prev_attr, active, params, events):
+        filtered_params = {k: v for k, v in params.items() if k not in self.credentialParams}
+        if prev_attr['active'] != active:
+            self._module.debug("active status differs")
+            return False
+        if self.hasCredentials(params):
+            self._module.debug("has credentials: assuming difference")
+            return False
+        if not self.eventsEqual(prev_attr, self.expandEvents(events)):
+            self._module.debug("events differs: %s != %s" % (prev_attr, events))
+            return False
+        if prev_attr['properties'] != filtered_params:
+            self._module.debug("services attributes differs: %s != %s" % (prev_attr['properties'], filtered_params))
+            return False
+        return True
+
+    def setattrs(self, obj, new_attr):
+        for k, v in new_attr.items():
+            setattr(obj, k, v)
+
+
+def main():
+    definitions = init_definitions()
+
+    base_specs = dict(
+        argument_spec=dict(
+            api_url=dict(required=True),
+            validate_certs=dict(required=False, default=True, type='bool'),
+            api_username=dict(required=False, no_log=True),
+            api_password=dict(required=False, no_log=True),
+            api_token=dict(required=False, no_log=True),
+            project=dict(required=True),
+            service=dict(required=False, type='str', choices=list(definitions.keys())),
+            active=dict(required=False, default=True, type='bool', choices=[True]),
+            params=dict(required=False, type='dict'),
```

Comment:
This argument violates our requirements for module arguments.  If you cannot enumerate all options this param takes, it violates our guidelines.

Please see item #5 at https://docs.ansible.com/ansible/latest/dev_guide/developing_modules.html

> Should you write multiple modules instead of one module?
> The functionality you want may be too large for a single module. If you want to connect Ansible to a new cloud provider, database, or network platform, you may need to develop a related group of modules.
> * Modules should have a concise and well defined functionality. Basically, follow the UNIX philosophy of doing one thing well.
> * Modules should not require that a user know all the underlying options of an API/tool to be used. For instance, if the legal values for a required module parameter cannot be documented, that’s a sign that the module would be rejected.
> * Modules should typically encompass much of the logic for interacting with a resource. A lightweight wrapper around an API that does not contain much logic would likely cause users to offload too much logic into a playbook, and for this reason the module would be rejected. Instead try creating multiple modules for interacting with smaller individual pieces of the API.

As such, this single module should become many modules that interact with individual components of the Gitlab API.

Reply:
+10k to what @sivel states above

Reply:
Even if we were to autogenerate 30 modules, I doubt they would be merged either.
See my comment below :  Best next option is to simply drop the validation of this option so that no validation = no listing = less quality but rules respected. (and user will take care of their parameters themselves)